File: common/src/main/java/com/cobblemontournament/common/api/tournament/TournamentProperties.java
package com.cobblemontournament.common.api.tournament;

import com.cobblemontournament.common.config.TournamentConfig;
import com.cobblemontournament.common.tournament.TournamentType;
import com.turtlehoarder.cobblemonchallenge.common.battle.ChallengeFormat;
import org.jetbrains.annotations.NotNull;

@NotNull
public final class TournamentProperties
{
    public static final TournamentType DEFAULT_TOURNAMENT_TYPE = TournamentConfig.DEFAULT_TOURNAMENT_TYPE;
    public static final int DEFAULT_GROUP_SIZE = TournamentConfig.DEFAULT_GROUP_SIZE;
    public static final int DEFAULT_MAX_PARTICIPANTS = TournamentConfig.DEFAULT_MAX_PARTICIPANTS;
    public static final ChallengeFormat DEFAULT_CHALLENGE_FORMAT = TournamentConfig.DEFAULT_CHALLENGE_FORMAT;
    public static final int DEFAULT_CHALLENGE_MIN_LEVEL = TournamentConfig.DEFAULT_MIN_LEVEL;
    public static final int DEFAULT_CHALLENGE_MAX_LEVEL = TournamentConfig.DEFAULT_MAX_LEVEL;
    public static final boolean DEFAULT_SHOW_PREVIEW = TournamentConfig.DEFAULT_SHOW_PREVIEW;

    public TournamentProperties()
    {
        tournamentType = DEFAULT_TOURNAMENT_TYPE;
        groupSize = DEFAULT_GROUP_SIZE;
        maxParticipants = DEFAULT_MAX_PARTICIPANTS;
        challengeFormat = DEFAULT_CHALLENGE_FORMAT;
        minLevel = DEFAULT_CHALLENGE_MIN_LEVEL;
        maxLevel = DEFAULT_CHALLENGE_MAX_LEVEL;
        showPreview = DEFAULT_SHOW_PREVIEW;
    }
    public TournamentProperties (
            TournamentType type,
            Integer groupSize,
            Integer maxParticipants,
            ChallengeFormat format,
            Integer minLevel,
            Integer maxLevel,
            Boolean showPreview
    ) {
        this.tournamentType = (type != null) ? type : DEFAULT_TOURNAMENT_TYPE;
        this.groupSize = groupSize != null ? groupSize : DEFAULT_GROUP_SIZE;
        this.maxParticipants = maxParticipants != null ? maxParticipants : DEFAULT_MAX_PARTICIPANTS;
        this.challengeFormat = format != null ? format : DEFAULT_CHALLENGE_FORMAT;
        this.minLevel = minLevel != null ? minLevel : DEFAULT_CHALLENGE_MIN_LEVEL;
        this.maxLevel = maxLevel != null ? maxLevel : DEFAULT_CHALLENGE_MAX_LEVEL;
        this.showPreview = showPreview != null ? showPreview : DEFAULT_SHOW_PREVIEW;
    }

    private final TournamentType tournamentType;
    private final Integer groupSize;
    private final Integer maxParticipants;
    private final ChallengeFormat challengeFormat;
    private final int minLevel;
    private final int maxLevel;
    private final boolean showPreview;

    @NotNull public TournamentType getTournamentType(){ return tournamentType; }
    public int getGroupSize(){ return groupSize; }
    public int getMaxPlayerCount(){ return maxParticipants; }
    @NotNull public ChallengeFormat getChallengeFormat(){ return challengeFormat; }
    public int getMinLevel(){ return minLevel; }
    public int getMaxLevel(){ return maxLevel; }
    public boolean getShowPreview(){ return showPreview; }
}

File: common/src/main/kotlin/com/cobblemontournament/common/tournament/properties/TournamentProperties.kt
package com.cobblemontournament.common.tournament.properties

import com.cobblemon.mod.common.api.reactive.Observable
import com.cobblemon.mod.common.api.reactive.ObservableSubscription
import com.cobblemon.mod.common.api.reactive.SettableObservable
import com.cobblemon.mod.common.api.reactive.SimpleObservable
import com.cobblemontournament.common.api.MatchMap
import com.cobblemontournament.common.api.MutableMatchMap
import com.cobblemontournament.common.api.MutablePlayerMap
import com.cobblemontournament.common.api.MutableRoundMap
import com.cobblemontournament.common.api.PlayerMap
import com.cobblemontournament.common.api.RoundMap
import com.cobblemontournament.common.api.cobblemonchallenge.ChallengeFormat
import com.cobblemontournament.common.config.TournamentConfig
import com.cobblemontournament.common.tournament.TournamentStatus
import com.cobblemontournament.common.tournament.TournamentType
import com.cobblemontournament.common.util.MATCH_MAP_HANDLER
import com.cobblemontournament.common.util.PLAYER_MAP_HANDLER
import com.cobblemontournament.common.util.ROUND_MAP_HANDLER
import com.sg8.collections.reactive.map.mutableObservableMapOf
import com.sg8.properties.DefaultProperties
import net.minecraft.nbt.CompoundTag
import net.minecraft.server.level.ServerPlayer
import java.util.UUID


class TournamentProperties(
    name: String = DEFAULT_TOURNAMENT_NAME,
    uuid: UUID = UUID.randomUUID(),
    tournamentStatus: TournamentStatus = DEFAULT_TOURNAMENT_STATUS,
    tournamentType: TournamentType = TournamentConfig.defaultTournamentType(),
    challengeFormat: ChallengeFormat = TournamentConfig.defaultChallengeFormat(),
    maxParticipants: Int = TournamentConfig.defaultMaxParticipants(),
    teamSize: Int = TournamentConfig.defaultTeamSize(),
    groupSize: Int = TournamentConfig.defaultGroupSize(),
    minLevel: Int = TournamentConfig.defaultMinLevel(),
    maxLevel: Int = TournamentConfig.defaultMaxLevel(),
    showPreview: Boolean = TournamentConfig.defaultShowPreview(),
    roundMap: RoundMap? = null,
    matchMap: MatchMap? = null,
    playerMap: PlayerMap? = null,
) : DefaultProperties<TournamentProperties> {

    override val instance: TournamentProperties = this
    override val helper = TournamentPropertiesHelper
    override val observable = SimpleObservable<TournamentProperties>()
    private val subscriptionsMap: MutableMap<Observable<*>, ObservableSubscription<*>> = mutableMapOf()

    private val _name = SettableObservable(name).subscribe()
    private val _uuid = SettableObservable(uuid).subscribe()
    private val _tournamentStatus = SettableObservable(tournamentStatus).subscribe()
    private val _tournamentType = SettableObservable(tournamentType).subscribe()
    private val _challengeFormat = SettableObservable(challengeFormat).subscribe()
    private val _maxParticipants = SettableObservable(maxParticipants).subscribe()
    private val _teamSize = SettableObservable(teamSize).subscribe()
    private val _groupSize = SettableObservable(groupSize).subscribe()
    private val _minLevel = SettableObservable(minLevel).subscribe()
    private val _maxLevel = SettableObservable(maxLevel).subscribe()
    private val _showPreview = SettableObservable(showPreview).subscribe()
    private var _roundMap: MutableRoundMap
    private var _matchMap: MutableMatchMap
    private var _playerMap: MutablePlayerMap

    var name: String
        get() = _name.get()
        set(value) { _name.set(value) }
    var uuid: UUID
        get() = _uuid.get()
        set(value) { _uuid.set(value) }
    var tournamentStatus: TournamentStatus
        get() = _tournamentStatus.get()
        set(value) { _tournamentStatus.set(value) }
    var tournamentType: TournamentType
        get() = _tournamentType.get()
        set(value) { _tournamentType.set(value) }
    var challengeFormat: ChallengeFormat
        get() = _challengeFormat.get()
        set(value) { _challengeFormat.set(value) }
    var maxParticipants: Int
        get() = _maxParticipants.get()
        set(value) { _maxParticipants.set(value) }
    var teamSize: Int
        get() = _teamSize.get()
        set(value) { _teamSize.set(value) }
    var groupSize: Int
        get() = _groupSize.get()
        set(value) { _groupSize.set(value) }
    var minLevel: Int
        get() = _minLevel.get()
        set(value) { _minLevel.set(value) }
    var maxLevel: Int
        get() = _maxLevel.get()
        set(value) { _maxLevel.set(value) }
    var showPreview: Boolean
        get() = _showPreview.get()
        set(value) { _showPreview.set(value) }
    var roundMap: MutableRoundMap
        get () = _roundMap
        set(value) { _roundMap = replaceSubscription(_roundMap, value) }
    var matchMap: MutableMatchMap
        get() = _matchMap
        set(value) { _matchMap = replaceSubscription(_matchMap, value) }
    var playerMap: MutablePlayerMap
        get() = _playerMap
        set(value) { _playerMap = replaceSubscription(_playerMap, value) }


    init {
        _roundMap = roundMap?.mutableCopy() ?: mutableObservableMapOf(ROUND_MAP_HANDLER)
        _roundMap.subscribe()
        _matchMap = matchMap?.mutableCopy() ?: mutableObservableMapOf(MATCH_MAP_HANDLER)
        _playerMap = playerMap?.mutableCopy() ?: mutableObservableMapOf(PLAYER_MAP_HANDLER)
    }


    private fun <T, O : Observable<T>> replaceSubscription(old: O, new: O): O {
        old.unsubscribe()
        return new.subscribe()
    }

    private fun <T, O : Observable<T>> O.subscribe(): O {
        subscriptionsMap[this] = this.subscribe { emitChange() }
        return this
    }

    private fun <T, O : Observable<T>> O.unsubscribe(): O {
        subscriptionsMap[this]?.unsubscribe()
        return this
    }

    override fun emitChange() = observable.emit(this)

    fun displaySlimInChat(player: ServerPlayer) {
        helper.displaySlimInChatHelper(properties = this, player = player)
    }

    fun displayResultsInChat(player: ServerPlayer) {
        helper.displayResultsInChatHelper(properties = this, player = player)
    }

    companion object {
        private val HELPER = TournamentPropertiesHelper
        fun loadFromNbt(nbt: CompoundTag) = HELPER.loadFromNbt(nbt = nbt)
    }

}
